com.unity.multiplayer.mlapi/Runtime/Core/MessageBatcher.cs

```diff
@@ -0,0 +1,198 @@
+            {
+                return read;
+            }
+            // otherwise, a two-byte length follows
+            int length = messageStream.ReadByte();
+            length += messageStream.ReadByte() * 256;
+
+            return length;
+        }
+
+        /// <summary>
+        /// FillTargetList
+        /// Fills a list with the ClientId's an item is targeted to
+        /// </summary>
+        /// <param name="item">the FrameQueueItem we want targets for</param>
+        /// <param name="list">the list to fill</param>
+        private static void FillTargetList(in FrameQueueItem item, ref List<ulong> list)
+        {
+            list.Clear();
+            switch (item.QueueItemType)
+            {
+                case RPCQueueManager.QueueItemType.ServerRpc:
+                    list.Add(item.NetworkId);
+                    break;
+                case RPCQueueManager.QueueItemType.ClientRpc:
+                    foreach (ulong clientId in item.ClientIds)
+                    {
+                        list.Add(clientId);
+                    }
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// QueueItem
+        /// Add a FrameQueueItem to be sent
+        /// </summary>queueItem
+        /// <param name="item">the threshold in bytes</param>
+        public void QueueItem(in FrameQueueItem item)
+        {
+            FillTargetList(item, ref TargetList);
+
+            foreach (ulong clientId in TargetList)
+            {
+                // todo: actually queue and buffer. For now, the dict contains just one entry !!!
+
+                if (!SendDict.ContainsKey(clientId))
+                {
+                    SendDict[clientId] = new SendStream();
+                    SendDict[clientId].Item = item;
+                    Writer.SetStream(SendDict[clientId].Stream);
+
+                    Writer.WriteBit(false); // Encrypted
+                    Writer.WriteBit(false); // Authenticated
+
+                    switch (item.QueueItemType)
+                    {
+                        // 6 bits are used for the message type, which is an MLAPIConstants
+                        case RPCQueueManager.QueueItemType.ServerRpc:
+                            Writer.WriteBits(MLAPIConstants.MLAPI_SERVER_RPC, 6); // MessageType
+                            break;
+                        case RPCQueueManager.QueueItemType.ClientRpc:
+                            Writer.WriteBits(MLAPIConstants.MLAPI_CLIENT_RPC, 6); // MessageType
+                            break;
+                    }
+                }
+
+                // write the amounts of bytes that are coming up
+                PushLength(item.MessageData.Count, ref Writer);
+
+                // write the message to send
+                // todo: is there a faster alternative to .ToArray()
+                Writer.WriteBytes(item.MessageData.ToArray(), item.MessageData.Count);
```

Comment:
I think you can also do item.Array.

Also, the type is IEnumerable so you could do your own little loop here and write a byte at a time to avoid the copy

Reply:
The MessageData is an ArraySegment. Doing `.Array` returns the whole underlying array, which is not what we want. The `ToArray()` call, here, copies just the correct section. I guess the todo and `.ToArray()` can stay for now. The explicit loop seems just contrite, and trading memory copies for cycles looping manually.